### DISCUSSION TITLE
How to scope mutate to a custom cache but not nested under the caches context provider?

### DISCUSSION THREAD

**Discussion by ssteuteville** [vercel/swr#1726] [Q&A]
Hi,

I'm a little bit confused on how to use mutate with a custom provider, outside of react. My situation is a bit unique, I'm building a module that ends up getting ran in an electron app with a custom javascript run time. Part of what I'm building requires a socket connection, so I'm trying to architect things as follows:

layer 1: Root level of app - Vanilla JS, should initialize cache data and subscribe to websockets and use mutate to keep the data fresh
layer 2: Root level renders 1 or more react apps
layer 3: components in react app use `useSWR` to get their data.

I want the apps in layer 2 to share cache, so i'm wrapping each in a `<SWRConfig value={{ provider: () => this.cache }}>`

The part that is unclear to me is, if I import `mutate` on layer 1, how do I make that mutate work with `this.cache` instead of the global provider?

Reading through your source code I see you guys are doing,

`const [cache, mutate] = initCache(....`

That seems like the affect I want but as far as I can tell nothing like that is exported. I also didn't see any options to configure a custom mutate for `<SWRConfig` to use.

Am I missing something or is there a way to use a mutate function that is scoped to a custom cache but not nested under the custom cache's context provider?

**Comment by ssteuteville** [comment]
1 solution that works for now is to use the global cache instead of one of my own caches, I'm worried I may need more control in the future though, so would still be interested in options!

**Comment by shuding** [comment]
We don’t expose the ability to mutate a scoped cache (SWRConfig’s provider) because of concerns regarding concurrent rendering. However we’re still exploring the safe ways to do so.
